onnxruntime/core/optimizer/matmul_scale_fusion.cc

```diff
@@ -0,0 +1,265 @@
+    float total_scale =
+        alpha_attr.has_type() && alpha_attr.type() == ONNX_NAMESPACE::AttributeProto_AttributeType_FLOAT
+            ? alpha_attr.f()
+            : 1.0f;
+    auto accumulate_scale = [&total_scale](const ScaleMergeInfo& fusion) {
+      total_scale *= fusion.scale;
+    };
+
+    std::for_each(input_node_merges.begin(), input_node_merges.end(), accumulate_scale);
+    std::for_each(output_node_merges.begin(), output_node_merges.end(), accumulate_scale);
+
+    alpha_attr = ONNX_NAMESPACE::MakeAttribute("alpha", total_scale);
+  }
+
+  auto get_mutable_node_to_merge = [&graph](const ScaleMergeInfo& merge) -> Node& {
+    return *graph.GetNode(merge.node_to_merge_edge->GetNode().Index());
+  };
+
+  std::vector<NodeArg*> fused_node_inputs = node.MutableInputDefs();
+  for (const auto& input_node_merge : input_node_merges) {
+    Node& input_node = get_mutable_node_to_merge(input_node_merge);
+    fused_node_inputs[input_node_merge.fused_node_def_index] =
+        input_node.MutableInputDefs()[input_node_merge.node_to_merge_def_index];
+  }
+
+  std::vector<NodeArg*> fused_node_outputs = node.MutableOutputDefs();
+  for (const auto& output_node_merge : output_node_merges) {
+    Node& output_node = get_mutable_node_to_merge(output_node_merge);
+    fused_node_outputs[output_node_merge.fused_node_def_index] =
+        output_node.MutableOutputDefs()[output_node_merge.node_to_merge_def_index];
+  }
+
+  Node& matmul_scale_node = graph.AddNode(
+      graph.GenerateNodeName("MatMul_With_Scale"),
+      "TransposeScaleMatMul",
+      "Fused MatMul and Scale",
+      fused_node_inputs,
+      fused_node_outputs,
+      &fused_node_attrs,
+      kMSDomain);
+
+  matmul_scale_node.SetExecutionProviderType(node.GetExecutionProviderType());
+
+  {
+    std::vector<std::reference_wrapper<Node>> nodes_to_remove{node};
+
+    for (const auto& input_node_merge : input_node_merges) {
+      // remove merged input node's output edge
+      auto input_node_edge = input_node_merge.node_to_merge_edge;
+      Node& input_node = get_mutable_node_to_merge(input_node_merge);
+      graph.RemoveEdge(
+          input_node.Index(), node.Index(),
+          input_node_edge->GetSrcArgIndex(), input_node_edge->GetDstArgIndex());
+
+      // only remove merged input node if it has no more outputs
+      if (!optimizer_utils::CheckOutputEdges(graph, input_node, 0)) continue;
+
+      nodes_to_remove.push_back(input_node);
+    }
+
+    for (const auto& output_node_merge : output_node_merges) {
+      nodes_to_remove.push_back(get_mutable_node_to_merge(output_node_merge));
+    }
+
+    for (Node& node_to_remove : nodes_to_remove) {
+      graph_utils::RemoveNodeOutputEdges(graph, node_to_remove);
+      graph.RemoveNode(node_to_remove.Index());
+    }
+  }
+
+  modified = true;
+
+  return Status::OK();
+}
+
+}  // namespace
+
+Status MatMulScaleFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger)
```

Comment:
It occurs to me that there is another condition we need to check: the fused scalar needs to be non-trainable constant, which doesn't require gradient. Otherwise, even if the Mul/Div has a initializer, we can't fuse them. 

See constant_folding.cc, excluded_initializers is a list of trainable-weights that we shouldn't fuse.